libgalois/src/analytics/connected_components/connected_components.cpp

```diff
@@ -0,0 +1,1341 @@
+    auto& me = graph.template GetData<NodeComponent>(n);
+    for (auto ii : graph.edges(n)) {
+      auto dest = graph.GetEdgeDest(ii);
+      auto& data = graph.template GetData<NodeComponent>(dest);
+      if (data != me) {
+        galois::gInfo(
+            "not in same component: ", (unsigned int)n, " (", me, ")", " and ",
+            (unsigned int)(*dest), " (", data, ")");
+        return true;
+      }
+    }
+    return false;
+  };
+
+  return galois::ParallelSTL::find_if(graph.begin(), graph.end(), is_bad) ==
+         graph.end();
+}
+
+galois::Result<ConnectedComponentsStatistics>
+galois::analytics::ConnectedComponentsStatistics::Compute(
+    galois::graphs::PropertyFileGraph* pfg, const std::string& property_name) {
+  using ComponentType = uint64_t;
+  struct NodeComponent : public galois::PODProperty<ComponentType> {};
+
+  using NodeData = std::tuple<NodeComponent>;
+  using EdgeData = std::tuple<>;
+  typedef galois::graphs::PropertyGraph<NodeData, EdgeData> Graph;
+  typedef typename Graph::Node GNode;
+
+  auto pg_result = Graph::Make(pfg, {property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+
+  auto graph = pg_result.value();
+
+  using Map = galois::gstl::Map<ComponentType, int>;
+
+  auto reduce = [](Map& lhs, Map&& rhs) -> Map& {
+    Map v{std::move(rhs)};
+
+    for (auto& kv : v) {
+      if (lhs.count(kv.first) == 0) {
+        lhs[kv.first] = 0;
+      }
+      lhs[kv.first] += kv.second;
+    }
+
+    return lhs;
+  };
+
+  auto mapIdentity = []() { return Map(); };
+
+  auto accumMap = galois::make_reducible(reduce, mapIdentity);
+
+  galois::GAccumulator<size_t> accumReps;
+
+  galois::do_all(
+      galois::iterate(graph),
+      [&](const GNode& x) {
+        auto& n = graph.template GetData<NodeComponent>(x);
+        accumMap.update(Map{std::make_pair(n, 1)});
+      },
+      galois::loopname("CountLargest"));
+
+  Map& map = accumMap.reduce();
+  size_t reps = map.size();
+
+  using ComponentSizePair = std::pair<ComponentType, int>;
+
+  auto sizeMax = [](const ComponentSizePair& a, const ComponentSizePair& b) {
+    if (a.second > b.second) {
+      return a;
+    }
+    return b;
+  };
+
+  auto identity = []() { return ComponentSizePair{}; };
+
+  auto maxComp = galois::make_reducible(sizeMax, identity);
+
+  galois::do_all(galois::iterate(map), [&](const ComponentSizePair& x) {
+    maxComp.update(x);
+  });
+
+  ComponentSizePair largest = maxComp.reduce();
+
+  // Compensate for dropping representative node of components
+  double ratio_largest_component = graph.size() - reps + map.size();
+  size_t largest_component_size = largest.second + 1;
+  if (ratio_largest_component) {
+    ratio_largest_component = largest_component_size / ratio_largest_component;
+  }
+
+  galois::GAccumulator<uint64_t> non_trivial_components;
+  galois::do_all(galois::iterate(map), [&](const ComponentSizePair& x) {
+    if (x.second > 1) {
+      non_trivial_components += 1;
+    }
+  });
```

Comment:
Maybe combine the two `do_all`s? Save some communication. Not that important though.

Reply:
Good catch.